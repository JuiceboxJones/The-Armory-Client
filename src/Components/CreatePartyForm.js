import React, { useState } from "react";

export default function CreatePartyForm(props) {
  const [partyName, setPartyname] = useState("");
  const [partyDescription, setPartyDescription] = useState("");
  const [chooseGame, setChooseGame] = useState("");
  const [partyRequirements, setPartyRequirements] = useState("");

  async function onPartyCreate() {
    try {
      // insert API services for party creation
      props.history.replace("/homepage?");
    } catch (error) {
      console.log(error.message);
    }
  }

  // generate party requirements page
  // generatePartyRequirements = () => {
  //   grab game.id and populate party requirements dropdown
  // };

  return (
    <main>
      <div className="createPartyForm">
        <h2>Create Party</h2>

        <form className="create-party-form" onSubmit={e => e.preventDefault()}>
          <div className="input-field">
            <label htmlFor="party-name-input">Party Name</label>
            <input
              id="party-name-input"
              type="text"
              placeholder="party"
              value={partyName}
              required
              onChange={e => setPartyname(e.target.value)}
            /><br/>
            <label htmlFor="party-description-input">Description</label>
            <input
              id="party-description-input"
              type="text"
              placeholder="limit 140 chars"
              value={partyDescription}
              required
              onChange={e => setPartyDescription(e.target.value)}
            /><br/>
            <label htmlFor="party-game-dropdown">Choose Game</label>
            <select
              id="party-game-dropdown"
              onChange={e => setChooseGame(e.target.value)}
              value={chooseGame}
            >
              <option>Overwatch</option>
              <option>Final Fantasy XIV</option>
              <option>League of Legends</option>
              <option>Fortnite</option>
              <option>Apex Legends</option>
              <option>DOTA 2</option>
              <option>CSGO</option>
              <option>Rainbow 6 Siege</option>
            </select><br/>
            <label htmlFor="party-requirement-input">Party Requirements</label>
            <select
              id="party-requirement-dropdown"
              onChange={e => setPartyRequirements(e.target.value)}
              value={partyRequirements}
            >
              {/* {generatePartyRequirements()} */}
              <option>Ranked</option>
              <option>Casual</option>
              <option>Competitive</option>
              <option>Fortnite</option>
            </select>
          </div>

          <button
            type="submit"
            className="create-party-button"
            onClick={onPartyCreate}
          >
            Create Party
          </button>
        </form>
      </div>
    </main>
  );
}
