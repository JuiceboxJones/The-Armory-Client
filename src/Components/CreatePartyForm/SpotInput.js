import React, { useState } from "react";
import AuthApiService from "../../services/auth-api-service";
import TokenService from "../../services/token-service";
import './SpotInput.css';

export default function SpotInput(props) {
  return (
    <div className={props.omitted ? 'spot_input spot_omitted' : 'spot_input'}>
      <button type="button" onClick={() => props.toggleOptions(props.index)}>
        {'Show Options' + props.index}
      </button>
      {props.showOptions && <div className="show_input_options">
        <button type="button" onClick={(e) => { props.omitSpot(props.index) }}> Omit </button>
      </div>}
    </div>
  );
}
