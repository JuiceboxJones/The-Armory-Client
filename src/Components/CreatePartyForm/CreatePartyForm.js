import React, { useState, useContext, useEffect } from "react";
import SpotInput from "./SpotInput/SpotInput";
import GameContext from '../../Contexts/gameContext';
import config from "../../config";
import TokenService from "../../services/token-service";
import Dropdown from '../Dropdown/Dropdown';

export default function CreatePartyForm(props) {
  const gameContext = useContext(GameContext);
  //inputs
  const [partyName, setPartyname] = useState("");
  const [partyDescription, setPartyDescription] = useState('');
  const [partyRequirement, setPartyRequirement] = useState(undefined);
  const [partyRequirement2, setPartyRequirement2] = useState(undefined);
  const [partyGamemode, setPartyGamemode] = useState(0);
  //spots
  const [spotMenuShown, setSpotMenuShown] = useState(undefined);
  const [spotsObj, setSpotsObj] = useState({});

  useEffect(() => {
    let temp = {};
    let blankSpot = {
      filled: null,
      roles: [],
      omitted: false,
      showOptions: false,
    };
    //we don't include the first one as a spot-input since the owner will always be included
    for (let i = 1; i < gameContext.partyLimit; i++) {
      temp[i] = blankSpot;
    }
    setSpotsObj(temp);
  }, []);

  function getActiveValues(target) {
    const spots = [];
    Object.entries(spotsObj).forEach(([_, value]) => {
      !value.omitted && spots.push(
        {
          filled: value.filled,
          roles: value.roles,
        }
      );
    });
    let gamemode = target.querySelector('button[name=gamemode]');
    gamemode ? (gamemode = gamemode.dataset.value) : (gamemode = 0);

    const [req, req2] = target.querySelectorAll('button[name=requirements]');
    const reqs = [];
    req && reqs.push(req.dataset.value);
    req2 && reqs.push(req2.dataset.value);

    return {
      room_id: props.roomUrl,
      party: {
        game_id: gameContext.id,
        title: partyName,
        description: partyDescription,
        gamemode: parseInt(gamemode),
        require_app: false,
      },
      spots,
      requirement: reqs,
    };
  }


  function onPartyCreate(e) {
    e.preventDefault();
    const newParty = getActiveValues(e.target);
    fetch(
        `${config.API_ENDPOINT}/parties`, 
        {
          method: 'POST',
          headers: {
            authorization: `Bearer ${TokenService.getAuthToken()}`,
            'content-type': 'application/json'
          },
          body: JSON.stringify(newParty),
        }
      )
      .then(res =>
        (!res.ok)
          ? res.json().then(e => Promise.reject(e))
          : res.json()
      )
      .then((respJson) => {
        props.history.push(`/party/${respJson}`);
      })
      .catch(err => {
        //UPDATE TO DISPLAY ERROR
        gameContext.setError(err);
      });
  }

  //SPOT HANDLERS
  function toggleSpotOptionsMenu(index) {
    (spotMenuShown === index)
      ? setSpotMenuShown(undefined)
      : setSpotMenuShown(index);
  }

  function toggleOmitSpot(index) {
    let newSpotsObj = {...spotsObj};
    newSpotsObj[index] = {
      ...newSpotsObj[index], 
      omitted: !newSpotsObj[index].omitted, 
      roles: [] 
    };
    setSpotsObj(newSpotsObj);
  }

  function handleSpotSubmit(index, roles) {
    let newSpotsObj = {...spotsObj};
    let newObj = {...newSpotsObj[index]};

    newObj.omitted = false;
    newObj.roles = roles;

    newSpotsObj[index] = newObj;
    setSpotsObj(newSpotsObj);
  }

  function generateSpotInputs() {
    let temp = [];
    let tempOmitted = [];

    Object.entries(spotsObj).forEach(([key, value]) => {
      let tempSpot = (
        <SpotInput
          key={key}
          index={key}
          omitted={value.omitted}
          toggleOmitSpot={toggleOmitSpot}
          toggleSpotOptionsMenu={toggleSpotOptionsMenu}
          showOptions={key === spotMenuShown}
          roles={value.roles}
          handleSpotSubmit={handleSpotSubmit}
        />
      );
      (value.omitted)
        ? tempOmitted.push(tempSpot)
        : temp.push(tempSpot);
    });

    return temp.concat(tempOmitted);
  }

  function getPartyRequirement2Options() {
    let temp = {...gameContext.requirements};
    delete temp[partyRequirement];

    return temp;
  }

  function handleReset(e) {
    setPartyname('');
    setPartyDescription('');
    setPartyRequirement(undefined);
    setPartyRequirement2(undefined);
    setPartyGamemode(0);
    const newSpotsObj = {...spotsObj};
    Object.keys(newSpotsObj).forEach(key => {
      newSpotsObj[key] = {...newSpotsObj[key], roles: [], omitted: false };
    });
    setSpotsObj(newSpotsObj);
    setSpotMenuShown(undefined);
  }

  return (
    <form className="create-party-form" onSubmit={onPartyCreate}>
      <h2>Create Party</h2>

        <div className="input-field">
          <label>Party Name
            <input
              type="text"
              placeholder="party"
              value={partyName}
              required
              onChange={e => setPartyname(e.target.value)}
            />
          </label>

          <label>Description
            <textarea
              id="party-description-input"
              type="text"
              maxLength='140'
              placeholder="limit 140 chars"
              value={partyDescription}
              onChange={e => setPartyDescription(e.target.value)}
            />
          </label>

          <fieldset className="create-party-form__filters">
            <legend>Requirements</legend>
            <Dropdown
              active={partyRequirement}
              name="requirements"
              onChange={e => setPartyRequirement(e.value)}
              onButtonClick={e => {
                if (partyRequirement2) {
                  setPartyRequirement(partyRequirement2);
                  setPartyRequirement2(undefined);
                } else {
                  setPartyRequirement(undefined);
                }
              }}
              startValue={partyRequirement}
              options={{...gameContext.requirements}}
            />
            <Dropdown
              active={partyRequirement2}
              inactive={!partyRequirement}
              onChange={e => setPartyRequirement2(e.value)}
              onButtonClick={() => setPartyRequirement2(undefined)}
              name="requirements"
              placeholder='Select a requirement...'
              startValue={partyRequirement2}
              options={getPartyRequirement2Options()}
            />
          </fieldset>
          <fieldset className="create-party-form__filters">
            <legend>Gamemode</legend>
            <Dropdown 
              active={partyGamemode}
              onChange={e => setPartyGamemode(e.value)}
              onButtonClick={() => setPartyGamemode(0)}
              startValue={partyGamemode}
              name="gamemode"
              gameId={gameContext.id}
              options={{...gameContext.gamemodes}}
            />
          </fieldset>
        </div>

        <div className='spot-container'>
          <div className="spot_input">Owner Spot</div>
          {generateSpotInputs()}
        </div>
        
        <button
          type="reset"
          className="create-party-reset-button"
          onClick={handleReset}
        >
          Reset
        </button>
        <button
          type="submit"
          className="create-party-button"
        >
          Create Party
        </button>
        <button
          aria-label="cancel"
          className="create-party-cancel-button"
          onClick={props.toggleCreatePartyForm}
        >
          Cancel
        </button>
    </form>
  );
}