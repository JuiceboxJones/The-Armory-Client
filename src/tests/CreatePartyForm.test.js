import React from 'react';
import CreatePartyForm from '../Components/CreatePartyForm/CreatePartyForm'
import ReactDOM from 'react-dom';
import { MemoryRouter } from 'react-router-dom';

describe('CreatePartyForm component', () => {

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <MemoryRouter>
        <CreatePartyForm />
      </MemoryRouter>, div
    );
    ReactDOM.unmountComponentAtNode(div);
  })
})