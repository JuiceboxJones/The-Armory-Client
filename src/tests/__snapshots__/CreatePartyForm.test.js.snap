// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreatePartyForm component renders the UI as expected 1`] = `
<form
  className="create-party-form"
  onSubmit={[Function]}
>
  <h2>
    Create Party
  </h2>
  <div
    className="input-field"
  >
    <label>
      Party Name
      <input
        onChange={[Function]}
        placeholder="party"
        required={true}
        type="text"
        value=""
      />
    </label>
    <label>
      Description
      <textarea
        id="party-description-input"
        maxLength="140"
        onChange={[Function]}
        placeholder="limit 140 chars"
        type="text"
        value=""
      />
    </label>
    <div
      className="dropdown"
    >
      <label>
        Party Requirements
        <div
          className="css-1pcexqc-container"
          onKeyDown={[Function]}
        >
          <div
            className="css-0 r-dropdown__control"
            onMouseDown={[Function]}
            onTouchEnd={[Function]}
          >
            <div
              className="css-1hwfws3 r-dropdown__value-container"
            >
              <div
                className="css-151xaom-placeholder r-dropdown__placeholder"
              >
                Select...
              </div>
              <input
                className="css-gj7qu5-dummyInput"
                disabled={false}
                id="react-select-4-input"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                readOnly={true}
                tabIndex="0"
                value=""
              />
            </div>
            <div
              className="css-1wy0on6 r-dropdown__indicators"
            >
              <span
                className="css-bgvzuu-indicatorSeparator r-dropdown__indicator-separator"
              />
              <div
                aria-hidden="true"
                className="css-16pqwjk-indicatorContainer r-dropdown__indicator r-dropdown__dropdown-indicator"
                onMouseDown={[Function]}
                onTouchEnd={[Function]}
              >
                <svg
                  aria-hidden="true"
                  className="css-19bqh2r"
                  focusable="false"
                  height={20}
                  viewBox="0 0 20 20"
                  width={20}
                >
                  <path
                    d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </label>
    </div>
    <div
      className="inactive-dropdown"
    />
    <div
      className="dropdown"
    >
      <label>
        Gamemode
        <div
          className="css-1pcexqc-container"
          onKeyDown={[Function]}
        >
          <div
            className="css-0 r-dropdown__control"
            onMouseDown={[Function]}
            onTouchEnd={[Function]}
          >
            <div
              className="css-1hwfws3 r-dropdown__value-container"
            >
              <div
                className="css-151xaom-placeholder r-dropdown__placeholder"
              >
                Select...
              </div>
              <input
                className="css-gj7qu5-dummyInput"
                disabled={false}
                id="react-select-5-input"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                readOnly={true}
                tabIndex="0"
                value=""
              />
            </div>
            <div
              className="css-1wy0on6 r-dropdown__indicators"
            >
              <span
                className="css-bgvzuu-indicatorSeparator r-dropdown__indicator-separator"
              />
              <div
                aria-hidden="true"
                className="css-16pqwjk-indicatorContainer r-dropdown__indicator r-dropdown__dropdown-indicator"
                onMouseDown={[Function]}
                onTouchEnd={[Function]}
              >
                <svg
                  aria-hidden="true"
                  className="css-19bqh2r"
                  focusable="false"
                  height={20}
                  viewBox="0 0 20 20"
                  width={20}
                >
                  <path
                    d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </label>
    </div>
  </div>
  <div
    className="spot-container"
  >
    <div
      className="spot_input"
    >
      Owner Spot
    </div>
  </div>
  <button
    className="create-party-reset-button"
    onClick={[Function]}
    type="reset"
  >
    Reset
  </button>
  <button
    className="create-party-button"
    type="submit"
  >
    Create Party
  </button>
  <button
    aria-label="cancel"
    className="create-party-cancel-button"
  >
    Cancel
  </button>
</form>
`;
